### PR TITLE
edit ndarray API docs

### PR DESCRIPTION
**Pull request by jiajiechen**
Edited some math function -- reduce function docs.
* nansum, nanprod, edited description. 
* sum, prod, mean, max, edited parameter description. 

Minor Edits
* ones, minor wording edits.

Edited some Array manipulation routines
* repeat, edit and add example.
* flip (reverse), edit and add example.

Edited some optimizer update function
* More edits in SGD update, momentum update, Adam update, edited description and parameters.
* For RMSprop and AlexRMSprop just edited parameters. 

@mli, @zackchase, @Roshrini, @madjam, @nswamy

### CODE REVIEW

File: src/operator/tensor/broadcast_reduce_op_value.cc [hunk 1]
```diff
@@ -15,18 +15,6 @@ DMLC_REGISTER_PARAMETER(BroadcastToParam);
 inline std::string get_reduce_axes_description(const std::string& op_name, int line) {
   std::string doc = R"code(Compute the __op__ of array elements over given axes.
 
-The argument ``axis`` specifies the axes to compute over:
```
Comment (piiswrong) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: why?

File: src/operator/tensor/broadcast_reduce_op_value.cc [hunk 2]
```diff
@@ -81,7 +69,7 @@ MXNET_OPERATOR_REGISTER_REDUCE_BACKWARD(_backward_nansum)
 .set_attr<FCompute>("FCompute<cpu>", ReduceAxesBackwardUseInOut<cpu, mshadow_op::nansum_grad>);
 
 MXNET_OPERATOR_REGISTER_REDUCE(nanprod)
-.describe(R"code(Compute the product of array elements over given axes with ``NaN`` ignored
+.describe(R"code(Compute the product of array elements over given axes treating Not a Numbers ``NaNs`` as zero.
 
 Refer to ``prod`` for more details.
```
Comment (nswamy) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: you can remove these, since the parameters now have good description.

File: src/operator/tensor/broadcast_reduce_op.h [hunk 3]
```diff
@@ -22,9 +22,14 @@ struct ReduceAxesParam : public dmlc::Parameter<ReduceAxesParam> {
   bool keepdims;
   DMLC_DECLARE_PARAMETER(ReduceAxesParam) {
     DMLC_DECLARE_FIELD(axis).set_default(TShape())
-        .describe("The axes to perform the reduction.");
+      .describe("The axis or axes along which to perform the reduction. "
+                "The default, `axis=()`, will compute over all elements into a "
+                "scalar array with shape `(1,)`. If axis is int, "
+                "a reduction is performed on a particular axis. "
+                "If axis is a tuple of ints, a reduction is performed "
+                "on all the axes specified in the tuple.");
```
Comment (nswamy) [inline src/operator/tensor/broadcast_reduce_op.h]: Can you find a way to make these different options appear on a new line.

File: src/operator/tensor/broadcast_reduce_op.h [hunk 4]
```diff
@@ -34,11 +39,12 @@ struct ReduceAxisParam : public dmlc::Parameter<ReduceAxisParam> {
   bool keepdims;
   DMLC_DECLARE_PARAMETER(ReduceAxisParam) {
     DMLC_DECLARE_FIELD(axis).set_default(dmlc::optional<int>())
-      .describe("int or None. The axis to perform the reduction. "
+      .describe("The axis along which to perform the reduction. "
                 "Negative values means indexing from right to left. "
```
Comment (nswamy) [inline src/operator/tensor/broadcast_reduce_op.h]: Do you need to mention this ? i do not see this mentioned for other parameters that have this behavior

File: src/operator/optimizer_op-inl.h [hunk 5]
```diff
@@ -212,24 +218,31 @@ struct RMSPropAlexParam : public dmlc::Parameter<RMSPropAlexParam> {
...
+    DMLC_DECLARE_FIELD(lr)
+    .describe("Learning rate");
+    DMLC_DECLARE_FIELD(gamma1).set_default(0.95f)
+    .describe("Decay rate.");
+    DMLC_DECLARE_FIELD(gamma2).set_default(0.9f)
+    .describe("Decay rate.");
+    DMLC_DECLARE_FIELD(epsilon).set_default(1e-8f)
+    .describe("A small constant for numerical stability.");
+    DMLC_DECLARE_FIELD(wd).set_default(0.0f)
+    .describe("Weight decay. "
```
Comment (zackchase) [inline src/operator/optimizer_op-inl.h]: Same as above:

Change to:

"Weight decay augments the objective function with a regularization term that penalizes large weights. The penalty scales with the square of the magnitude of each weight."

File: src/operator/optimizer_op-inl.h [hunk 6]
```diff
@@ -299,23 +312,29 @@ struct RMSPropParam : public dmlc::Parameter<RMSPropParam> {
...
-    DMLC_DECLARE_FIELD(epsilon).set_default(1e-8f).describe("epsilon");
-    DMLC_DECLARE_FIELD(wd).set_default(0.0f).describe("weight decay");
+    DMLC_DECLARE_FIELD(lr)
+    .describe("Learning rate");
+    DMLC_DECLARE_FIELD(gamma1).set_default(0.95f)
+    .describe("The dacay rate of momentum estimates.");
+    DMLC_DECLARE_FIELD(epsilon).set_default(1e-8f)
+    .describe("A small constant for numerical stability.");
+    DMLC_DECLARE_FIELD(wd).set_default(0.0f)
+    .describe("Weight decay. "
```
Comment (zackchase) [inline src/operator/optimizer_op-inl.h]: Change to:

"Weight decay augments the objective function with a regularization term that penalizes large weights. The penalty scales with the square of the magnitude of each weight."

File: src/operator/optimizer_op.cc [hunk 7]
```diff
@@ -16,19 +16,43 @@ DMLC_REGISTER_PARAMETER(RMSPropParam);
...
 .add_argument("weight", "NDArray-or-Symbol", "Weight")
-.add_argument("grad", "NDArray-or-Symbol", "gradient")
+.add_argument("grad", "NDArray-or-Symbol", "Gradient")
 .add_arguments(SGDParam::__FIELDS__());
 
 NNVM_REGISTER_OP(sgd_mom_update)
-.describe("Updater function for sgd optimizer")
+.describe(R"code(Momentum update function for Stochastic Gradient Descent (SDG) optimizer.
+
+Momentum update has better converge rates on neural networks. Mathematically it looks
```
Comment (zackchase) [inline src/operator/optimizer_op.cc]: Change "converge rates" to "convergence rates"

File: src/operator/tensor/broadcast_reduce_op_value.cc [hunk 8]
```diff
@@ -68,9 +56,7 @@ MXNET_OPERATOR_REGISTER_REDUCE_BACKWARD(_backward_prod)
 .set_attr<FCompute>("FCompute<cpu>", ReduceAxesBackwardUseInOut< cpu, mshadow_op::rdiv>);
 
 MXNET_OPERATOR_REGISTER_REDUCE(nansum)
-.describe(R"code(Compute the sum of array elements over given axes with ``NaN`` ignored
-
-Refer to ``sum`` for more details.
+.describe(R"code(Compute the sum of array elements over given axes treating Not a Numbers ``NaNs`` as zeros.
```
Comment (zackchase) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: the "s" cannot be inside the double backticks. Just put "treating each ``NaN`` as a zero."

File: src/operator/tensor/broadcast_reduce_op_value.cc [hunk 9]
```diff
@@ -81,9 +67,7 @@ MXNET_OPERATOR_REGISTER_REDUCE_BACKWARD(_backward_nansum)
 .set_attr<FCompute>("FCompute<cpu>", ReduceAxesBackwardUseInOut<cpu, mshadow_op::nansum_grad>);
 
 MXNET_OPERATOR_REGISTER_REDUCE(nanprod)
-.describe(R"code(Compute the product of array elements over given axes with ``NaN`` ignored
-
-Refer to ``prod`` for more details.
+.describe(R"code(Compute the product of array elements over given axes treating Not a Numbers ``NaNs`` as ones.
```
Comment (zackchase) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: Same as above

File: python/mxnet/ndarray.py [hunk 10]
```diff
@@ -953,7 +953,7 @@ def ones(shape, ctx=None, dtype=mx_real_t):
 
     Parameters
     ----------
-    shape : int or tuple of int
+    shape : int or sequence of ints
```
Comment (jiajiechen) [inline python/mxnet/ndarray.py]: Actually I took the numpy.ones as reference. 
https://docs.scipy.org/doc/numpy/reference/generated/numpy.ones.html
The `sequence` should represents both `list`, `tuple`.

File: src/operator/optimizer_op-inl.h [hunk 11]
```diff
@@ -27,18 +27,20 @@ struct SGDParam : public dmlc::Parameter<SGDParam> {
   float clip_gradient;
   DMLC_DECLARE_PARAMETER(SGDParam) {
     DMLC_DECLARE_FIELD(lr)
-    .describe("learning_rate");
+    .describe("Learning rate");
     DMLC_DECLARE_FIELD(wd)
     .set_default(0.0f)
-    .describe("weight decay");
+    .describe("Weight decay. "
+              "It is a regularization term that penalzies big weights, "
```
Comment (jiajiechen) [inline src/operator/optimizer_op-inl.h]: Good suggestion! Thanks!

**Comment by zackchase** [comment]
Numpy documentation is inconsistent. Please refer to
https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

If sequence is acceptable as a general word for {list, tuple} then I have
no problem. But should say "sequence of int" not "sequence of ints" to be
consistent with the *general* numpy style.

File: src/operator/tensor/matrix_op.cc [hunk 12]
```diff
@@ -423,24 +423,27 @@ NNVM_REGISTER_OP(_backward_clip)
...
 elements::
 
   x = [[ 1, 2],
        [ 3, 4]]
 
   repeat(x, repeats=2) = [ 1.,  1.,  2.,  2.,  3.,  3.,  4.,  4.]
 
-We can also choose a particular axis to repeat, in which a negative axis is
-interpreted counting from the backward::
+The parameter axis specifies the axis along which to perform repeat::
```
Comment (nswamy) [inline src/operator/tensor/matrix_op.cc]: The parameter `axis` specifies the axis along which to repeat::

File: src/operator/tensor/matrix_op.cc [hunk 13]
```diff
@@ -423,24 +423,27 @@ NNVM_REGISTER_OP(_backward_clip)
...
-  repeat(x, repeats=2, axis=-1) = [[ 1.,  2.],
-                                   [ 1.,  2.],
-                                   [ 3.,  4.],
-                                   [ 3.,  4.]]
+  repeat(x, repeats=2, axis=0) = [[ 1.,  2.],
+                                  [ 1.,  2.],
+                                  [ 3.,  4.],
+                                  [ 3.,  4.]]
+
+  repeat(x, repeats=2, axis=-1) = [[ 1.,  1.,  2.,  2.],
```
Comment (nswamy) [inline src/operator/tensor/matrix_op.cc]: you can say that the axis parameter supports negative indexing and remove this example

Comment (jiajiechen) [inline src/operator/tensor/matrix_op.cc]: I have said that in the parameter `axis` explanation.

File: src/operator/tensor/matrix_op.cc (hunk 12 above)
Comment (jiajiechen) [inline src/operator/tensor/matrix_op.cc]: fixed

File: src/operator/tensor/broadcast_reduce_op.h [hunk 14]
```diff
@@ -34,11 +39,12 @@ struct ReduceAxisParam : public dmlc::Parameter<ReduceAxisParam> {
...
-                "If is `None`, a global reduction will be performed.");
+                "``Requires axis to be set as int, because global reduction "
+                "is not supported yet.``");
```
Comment (piiswrong) [inline src/operator/tensor/broadcast_reduce_op.h]: This is used for multiple ops. Are all of them not supporting None?

Comment (Roshrini) [inline src/operator/tensor/broadcast_reduce_op.h]: I think this is for argmax,argmin and pick operator and these 3 don't support None yet.

File: src/operator/tensor/broadcast_reduce_op.h (hunk 4 above)
Comment (jiajiechen) [inline src/operator/tensor/broadcast_reduce_op.h]: what do you mean?

File: src/operator/tensor/broadcast_reduce_op.h (hunk 3 above)
Comment (jiajiechen) [inline src/operator/tensor/broadcast_reduce_op.h]: There is no easy way to spilt a new line.

File: src/operator/tensor/broadcast_reduce_op_value.cc (hunk 2 above)
Comment (jiajiechen) [inline src/operator/tensor/broadcast_reduce_op_value.cc]: what do you mean?

File: src/operator/optimizer_op-inl.h [hunk 15]
```diff
@@ -27,18 +27,20 @@ struct SGDParam : public dmlc::Parameter<SGDParam> {
   float clip_gradient;
   DMLC_DECLARE_PARAMETER(SGDParam) {
     DMLC_DECLARE_FIELD(lr)
-    .describe("learning_rate");
+    .describe("Learning rate");
```
Comment (jiajiechen) [inline src/operator/optimizer_op-inl.h]: they use .describe to document other parameters.

File: src/operator/tensor/matrix_op.cc (hunk 13 above)
Comment (jiajiechen) [inline src/operator/tensor/matrix_op.cc]: I think having a example for the negative indexing case might help understanding. what do you think? does it look redundant?

File: src/operator/tensor/broadcast_reduce_op.h (hunk 14 above)
Comment (jiajiechen) [inline src/operator/tensor/broadcast_reduce_op.h]: yes, it is only for argmax, argmin, and pick. These three does not support global reduction.

**Comment by zackchase** [comment]
what's the status on this @piiswrong ?

**Comment by piiswrong** [comment]
@zackchase waiting for approval. If you think this is ready to be merged I'll merge it

File: src/operator/tensor/broadcast_reduce_op.h (hunk 3 above)
Comment (piiswrong) [inline src/operator/tensor/broadcast_reduce_op.h]: you can put \n in the string.
